packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,198 @@
+		if ('error' in result) console.error(result.error)
+
+		if (Array.isArray(value)) console.table(cloneDeep(value))
+		else console.log(cloneDeep(value))
+
+		console.trace()
+
+		console.groupEnd()
+		/* eslint-enable no-console */
+
+		return false
+	}
+
+	// assign name for better debugging and to improve props rendering in the documentation
+	Object.defineProperty(validator, 'name', {
+		value: `propValidator('${key}')`,
+		writable: false,
+	})
+
+	return validator
+}
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+const walkShape = <SHAPE extends z.ZodTypeAny>(shape: SHAPE): Set<string> => {
+	/* eslint-disable no-console */
+	const types = new Set<string>()
+
+	let shapePointer = shape
+
+	while (shapePointer) {
+		const { typeName } = shapePointer._def
+
+		if (DEBUG_WALK_SHAPE)
+			console.log(
+				`walkTypes: found typeName “${typeName}” (${typeof typeName})`,
+			)
+
+		if (typeName === 'ZodUnion') {
+			if (DEBUG_WALK_SHAPE) console.group('walkTypes: walking union')
+
+			const options = shapePointer._def.options as Array<z.ZodTypeAny>
+
+			// walk every option in the enum and merge result sets into types
+			for (const typeSet of options.map((x) => walkShape(x)))
+				for (const type of typeSet) types.add(type)
+
+			if (DEBUG_WALK_SHAPE) console.groupEnd()
+		} else {
+			if (DEBUG_WALK_SHAPE) console.log(`walkTypes: pushing “${typeName}”`)
+			types.add(typeName)
+		}
+
+		shapePointer = shapePointer._def.innerType
+	}
+
+	return types
+	/* eslint-enable no-console */
+}
+
+/**
+ * These types do not influence the generation of Vue’s `propName.type` Arrays and are therefore discarded
+ */
+const ignoredZodTypes = new Set(['ZodDefault', 'ZodNullable', 'ZodOptional'])
+
+/**
+ * This maps the internal zod name of a type to the the constructor that Vue expects in `propName.type`
+ */
+const zodToVueTypes = new Map<
+	string,
+	Exclude<PropType<unknown>, Array<unknown>>
+>([
+	['ZodArray', Array],
+	['ZodBoolean', Boolean],
+	['ZodFunction', Function],
+	['ZodNativeEnum', String], // educated guess, can be fixed if the need for non-string enums arises
+	['ZodNumber', Number],
+	['ZodObject', Object],
+	['ZodString', String],
+])
+
+/**
+ * @summary Easily define & validate vue props, improve error messages drastically
+ *
+ * @description
+ * This augments/replaces Vue’s props entirely, including the need to add prop.validator
+ * by using zod and therefore actually prints errors to the console that can help developers
+ * rather than just say "hey you failed the validator, good luck!" as vue would do by default.
+ *
+ * ## Known limitations:
+ *
+ * 1. Deeply defined defaults are only applied when manually `schema.safeParse`-ing the data (only top-level defaults are possible to transform to Vue)
+ * 2. While this function also automatically types the return types via as PropType<> (which in theory
```

Comment:
via "as PropType<>"
or 
via asserting to PropType<>

Reply:
2. This function automatically types the return types of props using "as `PropType<>`" (which, in turn, makes the annotation of Vue props unnecessary). However, PropType<> appears to be bugged, as tested in @vue-composition/api (versions: 0.6.1 & 1.1.5).

Reply:
UP TO YOU